File: catalogizer-desktop/src/utils/__tests__/playbackFormatter.test.ts
import { describe, it, expect } from 'vitest';
import { PlaybackFormatter } from '../playbackFormatter';

describe('PlaybackFormatter', () => {
  it('formatAmount seconds → h/m/s', () => {
    expect(PlaybackFormatter.formatAmount(4980, 'seconds')).toBe('1h 23m');
    expect(PlaybackFormatter.formatAmount(1380, 'seconds')).toBe('23m');
    expect(PlaybackFormatter.formatAmount(45, 'seconds')).toBe('45s');
  });

  it('formatAmount pages singular/plural', () => {
    expect(PlaybackFormatter.formatAmount(1, 'pages')).toBe('1 page');
    expect(PlaybackFormatter.formatAmount(120, 'pages')).toBe('120 pages');
  });

  it('formatAmount events singular/plural', () => {
    expect(PlaybackFormatter.formatAmount(1, 'events')).toBe('1 run');
    expect(PlaybackFormatter.formatAmount(3, 'events')).toBe('3 runs');
  });

  it('formatAmount zero/negative → placeholder', () => {
    expect(PlaybackFormatter.formatAmount(0, 'seconds')).toBe('-');
    expect(PlaybackFormatter.formatAmount(-1, 'pages')).toBe('-');
  });

  it('formatProgress null/zero total falls back to current', () => {
    expect(PlaybackFormatter.formatProgress(1380, null, 'seconds')).toBe('23m');
    expect(PlaybackFormatter.formatProgress(1380, 0, 'seconds')).toBe('23m');
  });

  it('formatProgress renders slash form with total', () => {
    expect(PlaybackFormatter.formatProgress(1800, 7200, 'seconds')).toBe('30m / 2h 0m');
    expect(PlaybackFormatter.formatProgress(140, 320, 'pages')).toBe('140 pages / 320 pages');
  });

  it('progressFraction clamps 0..1', () => {
    expect(PlaybackFormatter.progressFraction(100, null)).toBe(0);
    expect(PlaybackFormatter.progressFraction(100, 0)).toBe(0);
    expect(PlaybackFormatter.progressFraction(50, 100)).toBe(0.5);
    expect(PlaybackFormatter.progressFraction(200, 100)).toBe(1);
  });

  it('formatReproductionCount × multiplier', () => {
    expect(PlaybackFormatter.formatReproductionCount(0)).toBe('0×');
    expect(PlaybackFormatter.formatReproductionCount(3)).toBe('3×');
  });
});

File: catalogizer-androidtv/app/src/test/java/com/catalogizer/androidtv/data/playback/PlaybackFormatterTest.kt
package com.catalogizer.androidtv.data.playback

import org.junit.Assert.assertEquals
import org.junit.Test

class PlaybackFormatterTest {

    @Test
    fun `formatAmount seconds renders hours and minutes`() {
        assertEquals("1h 23m", PlaybackFormatter.formatAmount(4980L, "seconds"))
        assertEquals("23m", PlaybackFormatter.formatAmount(1380L, "seconds"))
        assertEquals("45s", PlaybackFormatter.formatAmount(45L, "seconds"))
    }

    @Test
    fun `formatAmount pages uses singular for 1`() {
        assertEquals("1 page", PlaybackFormatter.formatAmount(1L, "pages"))
        assertEquals("120 pages", PlaybackFormatter.formatAmount(120L, "pages"))
    }

    @Test
    fun `formatAmount events uses singular for 1`() {
        assertEquals("1 run", PlaybackFormatter.formatAmount(1L, "events"))
        assertEquals("3 runs", PlaybackFormatter.formatAmount(3L, "events"))
    }

    @Test
    fun `formatAmount zero or negative returns placeholder`() {
        assertEquals("-", PlaybackFormatter.formatAmount(0L, "seconds"))
        assertEquals("-", PlaybackFormatter.formatAmount(-1L, "pages"))
    }

    @Test
    fun `formatProgress without total falls back to current`() {
        assertEquals("23m", PlaybackFormatter.formatProgress(1380L, null, "seconds"))
        assertEquals("23m", PlaybackFormatter.formatProgress(1380L, 0L, "seconds"))
    }

    @Test
    fun `formatProgress with total renders slash`() {
        assertEquals("30m / 2h 0m", PlaybackFormatter.formatProgress(1800L, 7200L, "seconds"))
        assertEquals("140 pages / 320 pages", PlaybackFormatter.formatProgress(140L, 320L, "pages"))
    }

    @Test
    fun `progressFraction clamps and handles null total`() {
        assertEquals(0f, PlaybackFormatter.progressFraction(100L, null), 0.001f)
        assertEquals(0f, PlaybackFormatter.progressFraction(100L, 0L), 0.001f)
        assertEquals(0.5f, PlaybackFormatter.progressFraction(50L, 100L), 0.001f)
        assertEquals(1f, PlaybackFormatter.progressFraction(200L, 100L), 0.001f)
    }

    @Test
    fun `formatReproductionCount appends multiplier`() {
        assertEquals("0×", PlaybackFormatter.formatReproductionCount(0L))
        assertEquals("3×", PlaybackFormatter.formatReproductionCount(3L))
    }
}
